airbyte-cdk/bulk/toolkits/load-azure-blob-storage/src/main/kotlin/io/airbyte/cdk/load/file/azureBlobStorage/AzureBlobStreamingUpload.kt

```
@@ -68,15 +65,19 @@ class AzureBlobStreamingUpload(
             } else {
                 val blockList = blockIds.values.toList()
                 log.info { "Committing block list for ${blockBlobClient.blobName}: $blockList" }
-                blockBlobClient.commitBlockList(blockIds.values.toList(), true) // Overwrite = true
             }
 
+            blockBlobClient.commitBlockList(blockIds.values.toList(), true) // Overwrite = true
+
             // Set any metadata
             if (metadata.isNotEmpty()) {
-                val filteredMetadata = filterInvalidMetadata(metadata)
-                if (filteredMetadata.isNotEmpty()) {
-                    blockBlobClient.setMetadata(filteredMetadata)
-                }
+                val mangledMetadata =
+                    metadata
+                        .map { (key, value) ->
+                            toValidMetadataKey(key) to toValidMetadataValue(value)
```

toValidMetadataValue might be sketchy? maybe better to just throw on invalid metadata - not entirely sure

---

I get anxious when I see BL leaking into utility classes like this. It kind of bakes into the client metadata and how we use it. I'd almost rather stick a metadataKey config value on the objectloader and inject it straight into the thing that manages the metadata

---

fair... I'll look into building a proper name mangler then. seems nicer than expecting people to hardcode a specific magic string in their connector.

(and I'll switch to throwing on invalid metadata values - seems safe for now, and mangling metadata values doesn't really provide a clear path to reading those values back anyway)

---

@johnny-schmidt how does the PR look now? I did some fairly naive plumbing to get the key mapper to all the places it needed to be, not sure if there's a better way to achieve that

in particular, I ended up injecting the key mapper into MSSQLWriter -> MSSQLBulkLoadStreamLoader, since they eventually call into PartToObjectAccumulator -> `ObjectStorageDestinationState.metadataFor(metadataKeyMapper, stream)`. Which feels a little roundabout.

---

ok, _now_ I think it's actually in proper shape for review

---

1. The streamloader code and PartToObjectAccumulator/RecordToPartAccumulator is dead old interface stuff. The relevant code is `ObjectLoader{PartFormatter|PartLoader|ObjectCompleter}`
2. Regardless, you shouldn't need to inject your metadata solution it into either code path. If you inject the metadata mapper into the destination state manager, then it can map whatever it provides from metadataFor (assuming you promote it to a non-static method).
3. You could also simplify it further by adding a metadataGenerationKey to ObjectLoader, and injecting a nullable ObjectLoader into DestinationState, and then override the key provided by metadataFor when ObjectLoader is present. That would be most targeted. (Like, right now this solution creates the maintenance burden of making sure anyone who uses metadata knows to inject a mapper. Whereas this solution would just literally replace the key in cases where the default one is incompatible.)